### PR TITLE
Feature #2745 mvmode enhancements

### LINKED ISSUE #1593: Fix Multivariate MODE to log it's output to a single log file.
## Describe the Enhancement ##
This task is closely related to MET #1283. If all Multivariate MODE functionality is run by a single process, then the logging issue effectively goes away.

MET #1184 was included in the met-10.0.0-beta2 release to add the multi-variate MODE option. This version of the code spawns new processes to execute individual calls to the standard MODE prior to combining the output into a super object. This task is to figure out how to handle the writing of a log file in this context. Since multiple processes are spawned, having them write to the same output log file is non-trivial.

Since the calls to MODE are done sequentially, consider adding a no-clobber option to the logger so that each call to single MODE can open and append to the same log file without overwriting it.

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
No sub-issues required.

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Review **projects** and select relevant **Repository** and **Organization** ones or add "alert:NEED PROJECT ASSIGNMENT" label
- [x] Select **milestone** to next major version milestone or "Future Versions"

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
No impacts.

## Enhancement Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)**, **Project(s)**, **Milestone**, and **Linked issues**
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### LINKED ISSUE #2707: Enhance Multivariate MODE to read input data only once rather than multiple times
*Replace italics below with details for this issue.*

## Describe the Enhancement ##
Current implementation reads data twice ('part1 simple' and 'part1 merge'.  Store in memory instead as part of part1 simple, access that in part1 merge. 

### Time Estimate ###
24 hours

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
January 16, 2024  (beta3)

### Funding Source ###
NOAA base

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [ ] Review default **alert** labels
- [ ] Select **component(s)**
- [ ] Select **priority**
- [ ] Select **requestor(s)**

### Milestone and Projects ###
- [ ] Select **Milestone** as the next official version or **Backlog of Development Ideas**
- [ ] For the next official version, select the **MET-X.Y.Z Development** project

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ]  [MET](https://github.com/dtcenter/MET/issues/new/choose), 

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the next official version
Select: **MET-X.Y.Z Development** project for development toward the next official release
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### LINKED ISSUE #2745: Major enhancements to multivariate MODE
## Describe the Enhancement ##
Several items did not make it into met12.0 beta2 that can be done for beta3:

### Time Estimate ###
3-5 days

### Sub-Issues ###
- [x] Add the original mvmode snow test case to the unit tests. Several configurations should be added.
- [x] Look further into the percentile thresholding to correctly implement more of the options and exclude those that cannot be done in mvmode.
- [x] Document the percentile threshold options that require both a forecast and an obs input, indicating that the indices must match in those cases.
- [x] Fix file_type so that it can be anywhere in the config, from innermost individual input to all the way out.
- [x] Consider multiple convolution thresholds and radii as an option for mvmode (currently disallowed).
- [x] Consider quilting as an option for mvmode (currently disallowed).
- [x] Add more checking for inconsistent/incorrect mvmode config settings. 
- [ ] Discuss mode with A.I., how would that be implemented (not anticipating actually doing anything besides discussing)

### Relevant Deadlines ###
beta3

### Funding Source ###
Split (for Dave):
2773543 NOAA BIL METp SW   40
2799991 Met Office METplus Support  40

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [x] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Review default **alert** labels
- [x] Select **component(s)**
- [x] Select **priority**
- [ ] Select **requestor(s)**

### Milestone and Projects ###
- [x] Select **Milestone** as the next official version or **Backlog of Development Ideas**
- [x] For the next official version, select the **MET-X.Y.Z Development** project

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose)

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [x] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [x] Complete the development and test your changes.
- [x] Add/update log messages for easier debugging.
- [x] Add/update unit tests.
- [x] Add/update documentation.
- [x] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the next official version
Select: **MET-X.Y.Z Development** project for development toward the next official release
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by davidalbo**
## Expected Differences ##

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[No]**</br>

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[No]**</br>

## Pull Request Testing ##

[ ] Describe testing already performed for these changes:   

- Tried configuring for various` file_type `settings, including invalid ones, placed both with individual inputs and common to all fcst or obs.  
- Did a variety of percentile threshold tests, including  `SFP, SOP` and `FBIAS` and in the debugger verified it was doing the right thing.  
- Ran through my existing suite of test runs, and compared the results to those from a run using the develop branch (because I'd made so many significant source code design changes).  </br>

- [ ] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:  I'd recommend playing around with configuring thresholding using percentiles, and configuring file_type, similar to what I did above.  You can find my test setup in `seneca:/d1/personal/dave/mvmode_test `with `run-multi.bsh` as a starting point.  The app built from this branch is there as` mode_new`.</br>

- [ ] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes  (after a few syntax errors!)]**

- [ ] Do these changes include sufficient testing updates? **[Yes]**

- [ ] Will this PR result in changes to the test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>
This branch has added new unit tests.  They are the tests equivalent to those in `/d1/personal/dave/mvmode_test`.  The input data has been added here:  ` seneca: /d1/projects/MET/MET_test_data/unit_test/model_data/mode_multivar ` to go with the unit test described here:  `internal/test_unit/xml/unit_mode_multivar.xml ` but I think the output that gets created by these new unit tests will break the testing, as those outputs are not yet added for comparison with 'truth'.  I don't quite know how to do that.

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR definition above.
- [ ] Ensure the PR title matches the feature or bugfix branch name.
- [ ] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the version that will include these changes
Select: **Coordinated METplus-X.Y Support** project for bugfix releases or **MET-X.Y.Z Development** project for official releases
- [ ] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Review by hertneky — APPROVED** [review]

**Review by hertneky — APPROVED** [review]

**Comment by JohnHalleyGotway** [comment]
@hertneky and @davidalbo, FYI, I just merged the latest from develop into this feature branch and compiled it on senenca in:
`/d1/projects/MET/MET_pull_requests/met-12.0.0/beta3/MET-feature_2745_mvmode_enhancements/bin`

We can run that version of MODE multivar to test this PR.

**Comment by hertneky** [comment]
@davidalbo I commit a UG change for multivar_name/level definitions, which I just noticed had old info.

**Review by hertneky — APPROVED** [review]
No issues found when testing this PR (Editing to add what was tested)
- running with normal thresholds w and w/o requesting intensities
- running various configurations of percentile thresholds SFP/SOP/FBIAS - noting expected errors such as FBIAS in both FCST/OBS, or FBIAS in FCST and SFP in OBS, etc

**Comment by davidalbo** [comment]
@JohnHalleyGotway I tried to resolve conflicts via command line, but it seems confusing (to me).  I cloned the repo, but then `git pull origin develop` did not work, it went south pretty bad.  I think I need to pull you in for guidance.

File: docs/Users_Guide/mode.rst [hunk 1]
```diff
@@ -118,6 +118,23 @@ The **multivar_intensity_compare_fcst** and **multivar_intensity_compare_obs** c
 
 When regridding to the FCST or OBS field (e.g. to_grid = FCST), the first field of the field array is used from the forecast and observation field dictionaries, respectively. All regridding is then done to that grid. Other regrid options described in :ref:`regrid` can also be used as normal.
 
+"file_type" can be set independently for each input in multivariate mode. If not set for an input, MET uses file names and file content to determine the type.
+
+When setting a threshold to a percentile, some choices require both an observation input and a forecast input.  When this is the case, it's assumed the indices match, so for example if forecast input 1 has such a percentile setting, then observation input 1 will be used to compute the percentile.  Percentiles in which this will happen are:
+
+* SFP in an observation input
```
Comment (JohnHalleyGotway) [inline docs/Users_Guide/mode.rst]:
```suggestion
* SFP in an observation input.
```

File: docs/Users_Guide/mode.rst [hunk 2]
```diff
@@ -118,6 +118,23 @@ The **multivar_intensity_compare_fcst** and **multivar_intensity_compare_obs** c
...
+  * The matching forecast input will be used to determine the threshold.  e.g. ">SFP33.3" in the 2nd observation input means greater than 33.3-rd percentile of the 2nd forecast input will be used as the threshold for that observation input.
+
+* SOP in a forecast input
```
Comment (JohnHalleyGotway) [inline docs/Users_Guide/mode.rst]:
```suggestion
* SOP in a forecast input.
```

**Review by JohnHalleyGotway — APPROVED** [review]
I approve of this pull request. I scrolled through the large set of file diffs.

I did push some changes to this feature for consistent whitespace handling. I also pushed Makefile changes cause by running `bootstrap` on seneca.

I don't fully grasp the intricate details of the changes needed to support percentile thresholding for multivar MODE. However, I'm glad that you added 3 new unit tests for MODE multivar and I'm glad that you and @hertneky iterated on testing these changes.

I reviewed the following:
- User's Guide doc updates.
- 3 new tests in `unit_mode_multivar.xml`
- Changes related to adding `clone` functions.
- Lots of changes in `mode_conf_info.cc` for handling percentile thresholds.
- Lots of changes in MODE and its frontends.

**Pull request merged**